File: 1577-number-of-ways-where-square-of-number-is-equal-to-product-of-two-numbers/1577-number-of-ways-where-square-of-number-is-equal-to-product-of-two-numbers.java
class Solution {
   public int numTriplets(int[] nums1, int[] nums2) {
    Map<Long, Long> map1 = new HashMap<>();
    for (int i = 0; i < nums1.length; i++)
        for (int j = i + 1; j < nums1.length; j++) {
            long prod = (long)nums1[i] * nums1[j];
            map1.put(prod, map1.getOrDefault(prod, 0l) + 1);
        }
    
    Map<Long, Long> map2 = new HashMap<>();
    for (int i = 0; i < nums2.length; i++)
        for (int j = i + 1; j < nums2.length; j++) {
            long prod = (long)nums2[i] * nums2[j];
            map2.put(prod, map2.getOrDefault(prod, 0l) + 1);
        }
    
    long res = 0l;
    for (long n : nums1) res += map2.getOrDefault((long)n * n, 0l);
    for (long n : nums2) res += map1.getOrDefault((long)n * n, 0l);
    
    return (int)res;
}
}

File: 1577-number-of-ways-where-square-of-number-is-equal-to-product-of-two-numbers/1577-number-of-ways-where-square-of-number-is-equal-to-product-of-two-numbers.cpp
class Solution {
public:
   int numTriplets(vector<int>& n1, vector<int>& n2) {
    return countForArray(n1, n2) + countForArray(n2, n1);
}
int countForArray(vector<int>& n1, vector<int>& n2) {
    int res = 0, last_res = 0, last_n = 0;
    sort(begin(n1), end(n1));
    for (auto i = 0; i < n1.size(); last_n = n1[i++])
        if (n1[i] == last_n)
            res += last_res;
        else
            res += last_res = twoProduct((long)n1[i] * n1[i], n2);
    return res;
}
int twoProduct(long i, vector<int> &n) {
    unordered_map<int, int> m;
    int cnt = 0;
    for (auto v : n) {
        if (i % v == 0)
            cnt += m[i / v];
        ++m[v];
    }
    return cnt;
} 
};